### PR TITLE
Add BadPacketsV

### PR DESCRIPTION
**Pull request by z7087**
Check for invalid finished digging packets
Should patch #1293 ~~#1296~~ but I only tested it on Spigot 1.8.8

### CODE REVIEW

**Comment by ManInMyVan** [comment]
Could also fix #1133

**Comment by z7087** [comment]
I think it is not enough as a fix for that

**Comment by ManInMyVan** [comment]
> I think it is not enough as a fix for that

Should atleast flag it tho

**Comment by z7087** [comment]
> > I think it is not enough as a fix for that
> 
> Should atleast flag it tho

Players can bypass it by send a start_digging packet before send the finished_digging packet that used for velocity

**Comment by SamB440** [comment]
Falses when breaking blocks in a protected region in my testing (Paper 1.20.4, 1.20.4 client)

**Comment by ghost** [comment]
> > > I think it is not enough as a fix for that
> > 
> > 
> > Should atleast flag it tho
> 
> Players can bypass it by send a start_digging packet before send the finished_digging packet that used for velocity

that going to flag fastbreak

**Comment by ManInMyVan** [comment]
> that going to flag fastbreak

99% of the time it mines air (check for breaking air?), I haven't tested it yet, but it may not flag when insta mining air.

**Comment by z7087** [comment]
> Falses when breaking blocks in a protected region in my testing (Paper 1.20.4, 1.20.4 client)

so #1296 can be vanilla :(
hope the Fastbreak check will get improved

**Comment by ghost** [comment]
> > Falses when breaking blocks in a protected region in my testing (Paper 1.20.4, 1.20.4 client)
> 
> so #1296 can be vanilla :( hope the Fastbreak check will get improved

if it that easy they already patch the velocity this way and not the workaround

File: src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                targetBlock = digging.getBlockPosition();
+            }
+
+            if (digging.getAction() == DiggingAction.FINISHED_DIGGING) {
+                if (targetBlock == null || !targetBlock.equals(digging.getBlockPosition())) {
+                    if (flagAndAlert("start: "+targetBlock+" finished: "+digging.getBlockPosition()) && shouldModifyPackets()) {
+                        event.setCancelled(true);
+                        player.onPacketCancel();
+
+                        // Ends the client prediction introduced in 1.19+
```
Comment (0-x-2-2) [inline src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java:44]: This is wrong and it is also wrong inside of Grim where you took this code from.

File: src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                        if (player.getClientVersion().isNewerThanOrEquals(ClientVersion.V_1_19) && PacketEvents.getAPI().getServerManager().getVersion().isNewerThanOrEquals(ServerVersion.V_1_19)) {
+                            player.user.sendPacket(new WrapperPlayServerAcknowledgeBlockChanges(digging.getSequence()));
+                        } else { // The client isn't smart enough to revert changes
```
Comment (0-x-2-2) [inline src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java:47]: The client is never smart enough to revert the changes like that even on 1.19+ if you don't sync blocks.

File: src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                            ResyncWorldUtil.resyncPosition(player, digging.getBlockPosition());
```
Comment (0-x-2-2) [inline src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java:48]: I wonder if this check should be moved to where Grim is already processing this data. There are possible exploits here with how little sanitization of these packets is done.

**Review by 0-x-2-2 — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java (hunk 1 above)
Comment (0-x-2-2) [reply src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java:44]: Screenshot

This is closer to the correct solution.

```java
ResyncWorldUtil.resyncPosition(player, pos);
FoliaCompatUtil.runTaskForEntity(player.bukkitPlayer, GrimAPI.INSTANCE.getPlugin(), () -> {
    // Ends the client prediction introduced in 1.19+
    if (player.getClientVersion().isNewerThanOrEquals(ClientVersion.V_1_19) && PacketEvents.getAPI().getServerManager().getVersion().isNewerThanOrEquals(ServerVersion.V_1_19)) {
        player.user.sendPacket(new WrapperPlayServerAcknowledgeBlockChanges(packet.getSequence()));
    }
}, null, 0);
```

Comment (SamB440) [reply src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java:44]: Is this what causes a block state desync that I'm seeing with Grim? (stand still, right click trapdoor in protected region that doesn't allow use, right click again, open state is desynced)

Comment (0-x-2-2) [reply src/main/java/ac/grim/grimac/checks/impl/badpackets/BadPacketsV.java:44]: Most likely! You can reproduce this easily with cheat clients + flagging a check, but it is possible to get that issue legitimately.

**Pull request closed without merging**